File: account-service/src/main/java/com/wks/servicemarketplace/accountservice/adapters/graphql/ApolloExecutionResult.java
package com.wks.servicemarketplace.accountservice.adapters.graphql;

import graphql.ExecutionResult;
import graphql.GraphQLError;

import java.util.List;
import java.util.Map;

/**
 * Apollo-server considers errors: [] to be an error.
 * This causes the service to not be detected.
 * As a workaround, ExecutionResult from GraphQL is wrapped in ApolloExecutionResult.
 * ApolloExecutionResult returns null when the errors array is empty.
 *
 * https://github.com/apollographql/apollo-client/issues/156
 */
public class ApolloExecutionResult implements ExecutionResult {

    private ExecutionResult executionResult;

    private ApolloExecutionResult(ExecutionResult executionResult) {
        this.executionResult = executionResult;
    }

    public static ExecutionResult of(ExecutionResult executionResult) {
        return new ApolloExecutionResult(executionResult);
    }

    @Override
    public List<GraphQLError> getErrors() {
        return executionResult == null
                || executionResult.getErrors() == null
                || executionResult.getErrors().isEmpty()
                ? null
                : executionResult.getErrors();
    }

    @Override
    public <T> T getData() {
        return executionResult == null ? null : executionResult.getData();
    }

    @Override
    public boolean isDataPresent() {
        return executionResult != null && executionResult.isDataPresent();
    }

    @Override
    public Map<Object, Object> getExtensions() {
        return executionResult == null ? null : executionResult.getExtensions();
    }

    @Override
    public Map<String, Object> toSpecification() {
        return executionResult == null ? null : executionResult.toSpecification();
    }
}

File: job-service/src/main/kotlin/com/wks/servicesmarketplace/jobservice/adapters/graphql/ApolloExecutionResult.kt
package com.wks.servicesmarketplace.jobservice.adapters.graphql

import com.fasterxml.jackson.annotation.JsonInclude
import graphql.ExecutionResult
import graphql.GraphQLError

/**
 * Apollo-server considers errors: [] to be an error.
 * This causes the service to not be detected.
 * As a workaround, ExecutionResult from GraphQL is wrapped in ApolloExecutionResult.
 * ApolloExecutionResult returns null when the errors array is empty.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
data class ApolloExecutionResult(private val executionResult: ExecutionResult?) : ExecutionResult {

    override fun toSpecification(): MutableMap<String, Any>? {
        if (this.executionResult?.toSpecification()?.isNotEmpty() == true) {
            return this.executionResult.toSpecification()
        }
        return null
    }

    override fun getErrors(): MutableList<GraphQLError>? {
        if (this.executionResult?.errors?.isNotEmpty() == true) {
            return this.executionResult.errors
        }
        return null
    }

    override fun isDataPresent(): Boolean {
        return this.executionResult?.isDataPresent == true
    }

    override fun getExtensions(): MutableMap<Any, Any>? {
        if (this.executionResult?.extensions?.isNotEmpty() == true) {
            return this.executionResult.extensions
        }
        return null
    }

    override fun <T : Any?> getData(): T? {
        return this.executionResult?.getData()
    }
}